libpromises/evalfunction.c

```diff
@@ -1188,6 +1188,95 @@ static FnCallResult FnCallGetGid(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const P
 #endif
 }
 
+/*********************************************************************/ 
+
+static FnCallResult FnCallUserInGroup(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    assert(fp != NULL);
+#ifdef _WIN32
+    Log(LOG_LEVEL_ERR, "Function '%s' is POSIX specific", fp->name);
+    return FnFailure();  
+#else
+    assert(finalargs != NULL);
+
+    const char *user_name = RlistScalarValue(finalargs);
+    const char *group_name = RlistScalarValue(finalargs->next);
+
+    StringSet *users = StringSetNew();
+    int ret;
+
+    // check secondary group
+    struct group grp;
+    struct group *grent;
+    char gr_buf[GETGR_R_SIZE_MAX] = {0};
+    ret = getgrnam_r(group_name, &grp, gr_buf, GETGR_R_SIZE_MAX, &grent);
+    
+    if (grent == NULL)
+    {
+        switch(ret)
+        {
+            case 0:
+            case ENOENT:
+            case ESRCH:
+            case EBADF:
+            case EPERM:
+                break;
+            default:
+            {
+                const char *error_msg = GetErrorStrFromCode(ret);
+                Log(LOG_LEVEL_ERR, "Couldn't open group database in function '%s': %s", fp->name, error_msg);
+                StringSetDestroy(users);
+                return FnFailure();
+            }
+        }
+    }
+    else {
+        while (grent->gr_mem[0] != NULL)
+        {
+            char *group_member = xstrdup(grent->gr_mem[0]);
+            StringSetAdd(users, group_member);
+            grent->gr_mem++;
+        }
+    }
+    
+    bool result = StringSetContains(users, user_name);
+    StringSetDestroy(users);
+
+    // check primary group
+    if (!result && grent != NULL)
+    {
+        struct passwd pwd;
+        struct passwd *pwent;
+        char pw_buf[GETPW_R_SIZE_MAX] = {0};
+        ret = getpwnam_r(user_name, &pwd, pw_buf, GETPW_R_SIZE_MAX, &pwent);
+
+        if (pwent == NULL)
```

Comment:
Same here, this if can be avoided with an early return.

Read this for some nice examples of this; https://refactoring.guru/replace-nested-conditional-with-guard-clauses